packages/grid/x-grid/src/tests/sorting.XGrid.test.tsx

```diff
@@ -221,5 +221,43 @@ describe('<XGrid /> - Sorting', () => {
       const time = Math.round(t1 - t0);
       expect(time).to.be.lessThan(300);
     });
+
+    it('should render maximum twice', async function test() {
+      // It's simpler to only run the performance test in a single controlled environment.
+      if (!/HeadlessChrome/.test(window.navigator.userAgent)) {
+        this.skip();
+        return;
+      }
+
+      let renderHeaderCount = 0;
+      const TestCasePerf = () => {
+        const [cols, setCols] = React.useState<GridColDef[]>([]);
+        const data = useData(5000, 10);
+
+        React.useEffect(() => {
+          if (data.columns.length) {
+            const newColumns = [...data.columns];
+            newColumns[1].renderHeader = (params) => {
+              renderHeaderCount += 1;
```

Comment:
As a general note, it would be better to move this into a useEffect, so we can leave React start and interrupts its rendering as he sees fit, and only see when a change is actually committed.

Reply:
Not sure what you mean.
If I add a useEffect in the TestComponent, I don't think it will be meaningful as it won't reflect the grid internal rendering

Reply:
In strict mode, render methods are called twice for each commit. In concurrent mode, React can render but not commit. In order to reliably test that memo is working, useEffect is a reliably to assert if a memo has worked, or not.

Reply:
@eps1lon Thanks for sharing these thoughts, however, I'm not sure we can act on it here.
The problem we are trying to solve is: how to reliably test that a memo logic is working (to avoid regressions)? I had linked https://github.com/mui-org/material-ui/blob/4a0e2459dc33c5e99df8419fc251db90638d65e7/packages/material-ui-styles/src/makeStyles/makeStyles.test.js#L508 as a way to approach it. What would you recommend?

Reply:
if you just want to add 
```tsx
        const rendered = React.useRef(1);
        React.useEffect(() => {
          rendered.current += 1;
        });
```
Then you need to add it in a component that will be rerendered when the grid renders, that could be a header, footer, cell, or column header. 

The way I did it, is via the `renderCell` and `renderHeader` functions, which get called when the cell or header render so it's the same. I can add an intermediary component such as below but I doubt it will add any value
```tsx
const TestCell = ()=> {
        const rendered = React.useRef(1);
        React.useEffect(() => {
          rendered.current += 1;
        });
...
}

renderCell: ()=> <TestCell {...params} />
```

Reply:
>  I can add an intermediary component such as below but I doubt it will add any value

Yes, this what I had assumed would ensure we reliably test the memo, based on: https://github.com/mui-org/material-ui-x/pull/1513/files#r629548477, e.g. if we are in strict mode, you should get two render call for each "interaction".

Reply:
> I'm not sure we can act on it here.

I was responding to your claims. Neither Damien nor I could make sense of them so I tried to bring some clarity.

> How to reliably test that a memo logic is working (to avoid regressions)?

You don't. You monitor performance because that's the feature `React.memo` is implementing. You don't test the actual implementation detail.